test/test_merkle.cpp

```diff
@@ -159,3 +162,313 @@ TORRENT_TEST(get_layer_offset)
+sha256_hash const gh = H(g,h);
+
+sha256_hash const ad = H(ab,cd);
+sha256_hash const eh = H(ef,gh);
+
+sha256_hash const ah = H(ad,eh);
+
+TORRENT_TEST(merkle_fill_tree)
+{
+	// fill whole tree
+	{
+		v tree{
+		       o,
+		   o,      o,
+		  o, o,   o, o,
+		a,b,c,d,e,f,g,h};
+
+		merkle_fill_tree(tree, 8, 7);
+
+		TEST_CHECK((tree ==
+		     v{
+		       ah,
+		   ad,     eh,
+		 ab, cd, ef, gh,
+		a,b,c,d,e,f,g,h}));
+	}
+
+	// fill left side of the tree
+	{
+		v tree{
+		       o,
+		   o,      o,
+		 ab,cd,   o, o,
+		a,b,c,d,o,o,o,o};
+
+		merkle_fill_tree(tree, 4, 7);
+
+		TEST_CHECK((tree ==
+		     v{o,
+		   ad,     o,
+		 ab, cd, o, o,
+		a,b,c,d,o,o,o,o}));
+	}
+
+	// fill right side of the tree
+	{
+		v tree{
+		       o,
+		   o,      o,
+		 o,  o,   o, o,
+		o,o,o,o,a,b,c,d};
+
+		merkle_fill_tree(tree, 4, 11);
+
+		TEST_CHECK((tree ==
+		     v{o,
+		   o,      ad,
+		 o,  o,  ab,cd,
+		o,o,o,o,a,b,c,d}));
+	}
+
+	// fill shallow left of the tree
+	{
+		v tree{
+		       o,
+		   o,      o,
+		 a,  b,   o, o,
+		o,o,o,o,o,o,o,o};
+
+		merkle_fill_tree(tree, 2, 3);
+
+		TEST_CHECK((tree ==
+		     v{o,
+		   ab,     o,
+		 a,  b,   o, o,
+		o,o,o,o,o,o,o,o}));
+	}
+
+	// fill shallow right of the tree
+	{
+		v tree{
+		       o,
+		   o,      o,
+		 o,  o,   a, b,
+		o,o,o,o,o,o,o,o};
+
+		merkle_fill_tree(tree, 2, 5);
+
+		TEST_CHECK((tree ==
+		     v{o,
+		   o,      ab,
+		 o,  o,   a, b,
+		o,o,o,o,o,o,o,o}));
+	}
+}
+
+TORRENT_TEST(merkle_root)
+{
+	// all leaves in the tree
+	TEST_CHECK(merkle_root(v{a,b,c,d,e,f,g,h}, o) == ah);
+
+	// not power-of-two number of leaves
+	TEST_CHECK(merkle_root(v{a,b,c,d,e,f}, o) == H(ad, H(ef, H(o, o))));
+
+	// very small tree
+	TEST_CHECK(merkle_root(v{a,b}, o) == ab);
+}
+
+namespace {
+void print_tree(span<sha256_hash const> tree)
+{
+	int const num_leafs = (tree.size() + 1) / 2;
+	int spacing = num_leafs;
+	int const num_levels = merkle_num_layers(num_leafs) + 1;
+	int layer_width = 1;
+	int node = 0;
+	for (int i = 0; i < num_levels; ++i)
+	{
+		for (int k = 0; k < layer_width; ++k)
+		{
+			for (int j = 0; j < spacing; ++j) std::cout << ' ';
+			std::cout << (tree[node] == sha256_hash() ? '0' : '1');
+			for (int j = 0; j < spacing - 1; ++j) std::cout << ' ';
+			++node;
+		}
+		std::cout << '\n';
+		layer_width *= 2;
+		spacing /= 2;
+	}
+	std::cout << '\n';
+}
+}
+
+TORRENT_TEST(merkle_clear_tree)
+{
+	// test clearing the whole tree
+	{
+		v tree{l,
+		   l,      l,
+		  l, l,   l, l,
+		l,l,l,l,l,l,l,l};
+
+		print_tree(tree);
+		merkle_clear_tree(tree, 8, 7);
+		print_tree(tree);
+
+		// notably, the root hash is left intact
+		TEST_CHECK((tree ==
+		     v{l,
+		   o,      o,
+		  o, o,   o, o,
+		o,o,o,o,o,o,o,o}));
+	}
+
+	// test clearing the left side of the tree
+	{
+		v tree{l,
+		   l,      l,
+		  l, l,   l, l,
+		l,l,l,l,l,l,l,l};
+
+		print_tree(tree);
+		merkle_clear_tree(tree, 4, 7);
+		print_tree(tree);
+
+		TEST_CHECK((tree ==
+		     v{l,
+		   l,      l,
```

Comment:
it is intentional to leave the root of the tree untouched, right?

Reply:
Yes that was intentional, although in hindsight not really necessary. The logic was that this would be used in places were the calculated root mismatches with the know valid root, so we wouldn't want to clear the valid root. However in practice I ended up clearing the tree first then replacing the root with the know valid value.

Reply:
Do you think this behavior should be changed?

Reply:
Yeah. Since it turned out to be unnecessary, leaving the root untouched is now just a potential gotcha for no good reason.